specs/local-context-spec.md

```diff
@@ -0,0 +1,274 @@
+- `clientContext.includeFiles` - OPTIONAL - a string of glob patterns for context files that MUST be included in addition to what may already be found. These paths MUST be relative to the AI working directory. e.g. `['additional-context/**', '/**/Desktop/shared-context/**']`
+- `clientContext.excludeFiles` - OPTIONAL - a string of glob patterns for context files that MUST NOT be included in the final set of context files. These paths MUST be relative to the AI working directory. e.g. `['**/private-context/**']`
+- `clientContext.ignoreGlobalContext` - OPTIONAL - a boolean that specifies if all global context should be ignored while doing work within the AI working directory. Default MUST be false.
+- `clientContext.ignoreAncestorContext` - OPTIONAL - a boolean that specifies if all ancestor context should be ignored while doing work within the AI working directory. Default MUST be false.
+
+Where `clientContext.includeFiles` and `clientContext.excludeFiles` overlap, the `excludeFiles` pattern takes precedence and any matching files MUST be excluded.
+
+When some or all fields are missing, the default configuration MUST apply the following values respectively:
+
+```json
+{
+  "clientContext": {
+    "includeFiles": [
+      "*"
+    ],
+    "excludeFiles": [
+      "context-config.json"
+    ],
+    "ignoreGlobalContext": false,
+    "ignoreAncestorContext": false
+  }
+}
+```
+
+AI systems MAY add more to the default configuration but MUST NOT conflict with the expected defaults above.
+
+AI systems MUST filter context file types that they do not support themselves and SHOULD show a warning message where there are selected context files that are not supported by the AI system.
+
+AI systems MUST validate the configuration file based on the above type signature and SHOULD show a warning message where there are configuration files that have invalid configuration. AI systems MUST NOT fail to start or work because of invalid configuration.
+
+### Merging algorithm for context and configuration
+
+The merging algorithm for context and configuration is as follows:
+1. source all context files and context configuration files based on the supported context scopes (global, static directory, dynamic subdirectory).
+2. create a list of absolute paths for context files and track which files are from which context scope.
+3. order all context configuration files based on the following precedence order where the closer to the AI working directory the configuration is, the higher priority it has:
+
+    1. global context configuration (least priority)
+    2. mounted directory context configuration
+    3. each ancestor context configuration that exists from leftmost directory to rightmost directory (most priority)
+4. recursively merge all context configuration files to generate the final configuration to apply
+    1. `clientContext.includeFiles` - arrays should be appended to and deduplicated based on pattern
```

Comment:
Are there any bounds for includeFiles globs? Is there anything preventing the AI system from traversing millions of files in node_modules for configuration?

Additionally, are there different rules for the global config?
For example, if `~/.context/context-config.json` contains:
```json
{
  "clientContext": {
    "includeFiles": ["**/.context/**"]
  }
}
```
does this mean that the AI system should traverse the entire filesystem in search for `.context` directories?